cfgov/data_research/models.py

```diff
@@ -11,3 +15,162 @@ class ConferenceRegistration(models.Model):
+    """
+    fips_type = models.CharField(max_length=6, default='county')
+
+
+class MSAMortgageData(MortgageBase):
+    """
+    A model for aggregating mortgage performance data from
+    a set of counties included in a metro area.
+    """
+    fips_type = models.CharField(max_length=6, default='msa')
+    counties = models.CharField(
+        max_length=255, blank=True, null=True,
+        help_text="A comma-separated list of FIPS for included counties.")
+    states = models.CharField(
+        max_length=255, blank=True, null=True,
+        help_text=("A comma-separated list of state abbreviations "
+                   "touched by FIPS for included counties."))
+
+    def save(self, **kwargs):
+        load_fips_meta()
+        self.aggregate_data()
+        if self.counties:
+            self.states = ", ".join(
+                sorted(set(
+                    [FIPS.county_fips.get(fips.strip())['state']
+                     for fips in self.counties.split(',')
+                     if FIPS.county_fips.get(fips.strip())])))
+        super(MSAMortgageData, self).save(**kwargs)
+
+    def aggregate_data(self):
+        if not self.counties:
+            return
+        count_fields = {
+            'total': 0, 'current': 0, 'thirty': 0,
+            'sixty': 0, 'ninety': 0, 'other': 0}
+        county_fips = [fips.strip() for fips in self.counties.split(',')]
+        county_records = CountyMortgageData.objects.filter(
+            fips__in=county_fips, date=self.date)
+        for county in county_records:
+            for field in count_fields:
+                count_fields[field] += getattr(county, field)
+        for field in count_fields:
+            setattr(self, field, count_fields[field])
+
+
+class StateMortgageData(MortgageBase):
+    """
+    A model to store aggregate state mortgage performance for a given date,
+    updated quarterly.
+    """
+
+    def save(self, **kwargs):
+        self.aggregate_data()
+        super(StateMortgageData, self).save(**kwargs)
+
+    def aggregate_data(self):
+        load_fips_meta()
+        state_fips_list = [fips for fips in FIPS.county_fips
+                           if fips[:2] == self.fips]
+        count_fields = {
+            'total': 0, 'current': 0, 'thirty': 0,
+            'sixty': 0, 'ninety': 0, 'other': 0}
+        county_records = CountyMortgageData.objects.filter(
+            date=self.date, fips__in=state_fips_list)
+        for county in county_records:
+            for field in count_fields:
+                count_fields[field] += getattr(county, field)
+        for field in count_fields:
+            setattr(self, field, count_fields[field])
+
+
+class NationalMortgageData(MortgageBase):
+    """
+    A model to store aggregate national mortgage performance for a given date,
+    updated quarterly.
+    """
+
+    def save(self, **kwargs):
+        self.aggregate_data()
+        super(NationalMortgageData, self).save(**kwargs)
+
+    def aggregate_data(self):
```

Comment:
I think I'd like to do this optimization after the self.aggregate_data methods are less fluid. The sequence of when self_aggregate_data is called might be changing.